libimage/push.go

```diff
@@ -29,28 +32,79 @@ func (r *Runtime) Push(ctx context.Context, source, destination string, options
 		options = &PushOptions{}
 	}
 
-	// Look up the local image.  Note that we need to ignore the platform
-	// and push what the user specified (containers/podman/issues/10344).
-	image, resolvedSource, err := r.LookupImage(source, nil)
+	// Push the single image
+	if !options.AllTags {
+
+		// Look up the local image.  Note that we need to ignore the platform
+		// and push what the user specified (containers/podman/issues/10344).
+		image, resolvedSource, err := r.LookupImage(source, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		// Make sure we have a proper destination, and parse it into an image
+		// reference for copying.
+		if destination == "" {
+			// Doing an ID check here is tempting but false positives (due
+			// to a short partial IDs) are more painful than false
+			// negatives.
+			destination = resolvedSource
+		}
+
+		return pushImage(ctx, image, destination, options, resolvedSource, r)
+	}
+
+	// Below handles the AllTags option, for which we have to build a list of
+	// all the local images that match the provided repository and then push them.
+	//
+	// Start by making sure a destination was not specified, since we'll get that from the source
+	if len(destination) != 0 {
+		return nil, fmt.Errorf("`destination` should not be specified if using AllTags")
+	}
+
+	// Make sure the source repository does not have a tag
+	srcNamed, err := reference.ParseNormalizedNamed(source)
 	if err != nil {
 		return nil, err
 	}
+	if _, hasTag := srcNamed.(reference.NamedTagged); hasTag {
+		return nil, fmt.Errorf("can't push with AllTags if source tag is specified")
+	}
 
+	// Now list every image of that source repository
+	listOptions := &ListImagesOptions{}
+	listOptions.Filters = []string{fmt.Sprintf("reference=%s:*", source)}
+	logrus.Debugf("Finding all images for source %s", source)
+	srcImages, _ := r.ListImages(ctx, nil, listOptions)
+
+	// Push each tag for every image in the list
+	var byteManifest []byte
+	for _, img := range srcImages {
+		namedTagged, err := img.NamedTaggedRepoTags()
+		if err != nil {
+			return nil, err
+		}
+		for _, n := range namedTagged {
+			destWithTag := fmt.Sprintf("%s:%s", source, n.Tag())
+			b, err := pushImage(ctx, img, destWithTag, options, "", r)
+			if err != nil {
+				return byteManifest, err
+			}
+			byteManifest = append(byteManifest, b...)
```

Comment:
I can’t see what the resulting value could be good for. Either punt and return `nil`, or… I don’t know. Return to the `PushAllTags` idea and return a list of the individual manifests? Even that is not really useful without also listing the repo:tag values associated with each manifest.

Reply:
I was trying to preserve the behavior of the underlying copy(), which returns the manifest.

Reply:
But tbh if they are just smashed together, I don't know if they are or any use. Like you are saying.

Reply:
- A concatenation of two manifests isn’t a manifest
- AFAIK the manifest is returned so that `podman push` can write the resulting digest to a file. That concept doesn’t _directly_ translate for pushing multiple images, everything in the call stack up to and including the script invoking the CLI needs adjusting for multiple results instead of a single one.

Reply:
I think it's OK to return `nil` if all is specified but we need to document that in the `AllTags` field and the `Push()` docs.  Alternatively, we could change the `[]byte` to a `[][]byte` but I am not sure how useful that is to callers who usually just use it to write a single digest to a file.

Reply:
Opted to return `nil`